### PR TITLE
Refactor sweep system to adaptive experiments framework


### PR DESCRIPTION
**Pull request by akrbc9**
Refactor sweep system to adaptive experiments framework

This PR introduces a new adaptive experiments system that generalizes the existing sweep functionality:

### New Adaptive Framework
- **AdaptiveController**: Core orchestration logic for running experiments
- **AdaptiveConfig**: Simplified configuration for adaptive experiments  
- **AdaptiveTool**: CLI tool for running adaptive experiments with multiple scheduler types

### Scheduler System
- **TrainAndEvalScheduler**: Simple scheduler for proof-of-concept train-then-eval workflows
- **BatchedSyncedScheduler**: Bayesian optimization scheduler (moved from sweep system)
- Pluggable scheduler architecture supporting different experiment types

### Infrastructure
- **LocalDispatcher & SkypilotDispatcher**: Job execution backends
- **WandbStore**: Experiment state persistence via Weights & Biases
- Unified job definition and run tracking models

### Sweep Tool Updates
- Refactored SweepTool to use adaptive framework under the hood
- Maintains backward compatibility for existing sweep workflows
- Enhanced Bayesian optimization with proper observation tracking

### Testing & Examples
- Comprehensive unit tests for schedulers and core logic
- Smoke tests for end-to-end workflow validation
- Example adaptive experiment recipes

## Migration Path
- Existing sweep functionality preserved through SweepTool
- New adaptive experiments available via AdaptiveTool
- Gradual migration path for complex experiment workflows

[Asana Task](https://app.asana.com/1/1209016784099267/project/1210348820405981/task/1211400976998724)

### CODE REVIEW

**Comment by akrbc9** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/Metta-AI/metta/2680?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#2680** <a href="https://app.graphite.dev/github/pr/Metta-AI/metta/2680?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/Metta-AI/metta/2680?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#2645** <a href="https://app.graphite.dev/github/pr/Metta-AI/metta/2645?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#2671](https://github.com/Metta-AI/metta/pull/2671) <a href="https://app.graphite.dev/github/pr/Metta-AI/metta/2671?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: smoke_tests/adaptive/README.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
+# Adaptive System Smoke Tests
+
```
Comment (akrbc9) [inline smoke_tests/adaptive/README.md:2]: may not be necessary

File: smoke_tests/adaptive/run_all.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
+#!/usr/bin/env python3
+"""Convenience script to run all adaptive smoke tests.
+
```
Comment (akrbc9) [inline smoke_tests/adaptive/run_all.py:3]: may not be necessary

File: tests/adaptive/test_train_and_eval_scheduler.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
+"""Tests for TrainAndEvalScheduler."""
+
```
Comment (akrbc9) [inline tests/adaptive/test_train_and_eval_scheduler.py:2]: remove

File: tests/sweep/test_batched_synced_scheduler_state.py [hunk 4]
```diff
@@ -0,0 +1,589 @@
+"""Unit tests for BatchedSyncedOptimizingScheduler state tracking.
+
+Tests all edge cases of internal state management including:
```
Comment (jacklionheart) [inline tests/sweep/test_batched_synced_scheduler_state.py:3]: In general I try to avoid directly testing internal state, as it tends to be the most brittle to refactoring.  I try as much as possible to do behavioral testing on external-facing APIs.

**Review by jacklionheart — APPROVED** [review]
I stand by most of my old comments, but we can hash it out post-merge
